File: src/text_subscriber.cpp
#include "ros/ros.h"
#include "performance_tests/SuperAwesome.h"
#include "performance_tests/RequestSubscriberStats.h"


// bool add(beginner_tutorials::AddTwoInts::Request  &req,
//          beginner_tutorials::AddTwoInts::Response &res)
// {
//   res.sum = req.a + req.b;
//   ROS_INFO("request: x=%ld, y=%ld", (long int)req.a, (long int)req.b);
//   ROS_INFO("sending back response: [%ld]", (long int)res.sum);
//   return true;
// }


class DataSubscriber
{
  private:
    int count;
    float sum;
    bool reset;
    ros::Time begin,end;
    ros::Duration time_lapse;
  public:
    float avg;
    DataSubscriber();
    void chatterCallback(const performance_tests::SuperAwesome::ConstPtr& msg);
    bool requestAverageRate(performance_tests::RequestSubscriberStats::Request  &req,
                             performance_tests::RequestSubscriberStats::Response  &res);
};

DataSubscriber::DataSubscriber():sum(0),count(0),reset(false){}

void DataSubscriber::chatterCallback(const performance_tests::SuperAwesome::ConstPtr& msg)
{
  end = ros::Time::now();
  ros::Duration time_lapse = end-begin;
  //ROS_INFO("%f", 1/time_lapse.toSec());
  if (reset == true)
  {
    sum = 0; count=0;
    reset = false;
  }
  else
  {  
    sum = sum + time_lapse.toSec();
    count = count + 1;
  }
  begin = ros::Time::now();
}


bool DataSubscriber::requestAverageRate(performance_tests::RequestSubscriberStats::Request  &req,
                                        performance_tests::RequestSubscriberStats::Response  &res)
{
 avg = 1/(sum/count); 
 res.avg_rate = avg;
 reset = req.reset; 
 return true;
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "text_subscriber_cpp"); 
  ros::NodeHandle n;
  DataSubscriber ds;
  ros::Subscriber sub = n.subscribe("performance_test_msg", 1000, &DataSubscriber::chatterCallback, &ds);
  ros::ServiceServer stats_service = n.advertiseService("request_subscriber_stats", &DataSubscriber::requestAverageRate, &ds);
  ros::spin();
  return 0;
}

File: src/text_subscriber.py
#!/usr/bin/env python
import rospy
from performance_tests.msg import SuperAwesome
from performance_tests.srv import *

begin = 0
end = 0
avg = 0
sum = 0
count = 0
reset = 0

def cb(data):
    global end,begin,sum,reset,count
    end = rospy.get_rostime()
    time_lapse = (end-begin).to_sec()
    # rospy.loginfo("%f", 1/time_lapse)
    if (reset == True):
        sum = 0; count=0;
        reset = False
    else: 
        sum = sum + time_lapse
        count = count + 1;
    begin = rospy.get_rostime()

def request_stats(req):
    global sum,reset
    reset = req.reset
    avg = 1/(sum/count); 
    return RequestSubscriberStatsResponse(avg)
        
    
if __name__ == '__main__':
    rospy.init_node('text_subscriber_py', anonymous=True)
    rospy.Subscriber("performance_test_msg", SuperAwesome, cb,queue_size=1000)
    s = rospy.Service('request_subscriber_stats', RequestSubscriberStats, request_stats)
    begin = rospy.get_rostime()
    rospy.spin()
